sonic-ycabled/ycable/ycable_utilities/y_cable_helper.py

```diff
@@ -4080,4 +4013,120 @@ def join(self):
  
             raise self.exc
 
+class GracefulRestartClient:
+    def __init__(self, port, channel: grpc.aio.secure_channel, read_side):
+        self.port = port
+        self.stub = linkmgr_grpc_driver_pb2_grpc.GracefulRestartStub(channel)
+        self.request_queue = asyncio.Queue()
+        self.response_queue = asyncio.Queue()
+        self.read_side = read_side
+
+    async def send_request_and_get_response(self):
+        while True:
+            tor = await self.request_queue.get()
+            request = linkmgr_grpc_driver_pb2.GracefulAdminRequest(tor=tor)
+            response = None 
+            try:
+                response_stream = self.stub.NotifyGracefulRestartStart(request)
+                index = 0
+                async for response in response_stream:
+                    helper_logger.log_notice("Async client received from direct read period port = {}: period = {} index = {} guid = {} notifytype {} msgtype = {}".format(self.port, response.period, index, response.guid, response.notifytype, response.msgtype))
+                    helper_logger.log_debug("Async Debug only :{} {}".format(dir(response_stream), dir(response)))
+                    index = index+1
+                    if response == grpc.aio.EOF:
+                        break
+                helper_logger.log_notice("Async client finished loop from direct read period port:{} ".format(self.port))
+                index = index+1
+            except grpc.RpcError as e:
+                helper_logger.log_notice("Async client port = {} exception occured because of {} ".format(self.port, e.code()))
+
+            await self.response_queue.put(response)
+
+    async def process_response(self):
+        while True:
+            response = await self.response_queue.get()
+            helper_logger.log_debug("Async recieved a response from {} {}".format(self.port, response))
+            # do something with response
+            if response is not None:
+                await asyncio.sleep(response.period)
+            else:
+                await asyncio.sleep(20)
```

Comment:
Will server respond when there isn't a graceful restart? 

How is 20s determined here? If ToR doesn't get graceful restart notification before the scheduled restart, will SoC go ahead and restart?

Reply:
Discussed offline, as mentioned 20seconds sleep is when there is no soc connectivity, otherwise once the notification for service start comes, we would sleep for advertised period.